Copied (and adapted) existing k-means test cases to new k-means

fschopp opened the pull request:
_This is still work in progress. Pull request opened to facilitate discussion. Please only merge when requested._

fschopp commented:
@huor It seems that `dataset.xml` and `kmeans.xml` (and probably the casespecs of other modules) contain redundant information. Is it really necessary to list a method in `dataset.xml`?

fschopp commented:
@huor Are there assumption on the names in `dataset.xml`, `algorithms.xml`, and the casespecs (such as `kmeans.xml`)? For instance, is a `test_suite` assumed to be named `<method>_<name>`?

huor commented:
The algorithms/methods defined in algorithms.xml, and datasets defined in dataset.xml will be used in kmeans.xml.
The algorithms/methods defined in algorithms.xml also can be used in dataset.xml.

huor commented:
The information in dataset.xml and kmeans.xml is not redundant. The the datasets defined dataset.xml is a kind of algorithms-prarameters combination, it will ease us to write cases in kmeans.xml

fschopp commented:
So, you are saying that, e.g., the fact that test suite `kmeans_cset_negative_src_relation` [contains `init_cset_rel`](https://github.com/madlib/testsuite/blob/6bc781c5f60dac8c3e518303e10e53b1edbcf0aa/testspec/casespec/kmeans.xml#L878) and that `dataset.xml` [contains the same parameter](https://github.com/madlib/testsuite/blob/6bc781c5f60dac8c3e518303e10e53b1edbcf0aa/testspec/metadata/dataset.xml#L6272) is an oversight and not necessary?

fschopp commented:
Regarding the naming conventions: Yes, I see that some names are references. But I am wondering if you also make assumption that names are composed of different substrings? Do you ever concatenate string and then assume that there is a certain algorithm/test suite/method with that composed name?

geeg commented on testspec/casespec/kmeans.xml:
```diff
@@ -101,6 +143,24 @@
+                </list_parameter>
+                <list_parameter>
+                    <name>dist_metric</name>
+                    <value>squared_dist_norm1</value><value>squared_dist_norm2</value><value>squared_angle</value><value>squared_tanimoto</value>
```
I found that I had to explicitly include the madlib schema, e.g.: <value>madlib.squared_dist_norm1</value>
(Similar change had to be made in the algorithmspec.xml file for madlib.avg)

geeg commented:
The MADmark Installation Guide that Jiali sent explains everything pretty well.  Here's a summary of how to run select test cases:

After updating the 3 XML files accordingly, run 'cd $MADMARK_HOME/bin; python run.py -g' to generate all your test cases.  So, if in kmeans.xml, you have a test_suite named "kmeans_baseline" and you have a total of 4 combinations of parameters to test, you'll be generating four case files in $MADMARK_HOME/testcase:
  kmeans_baseline_0_0
  kmeans_baseline_0_1
  kmeans_baseline_0_2
  kmeans_baseline_0_3

To run select test cases:
1) Create a .yaml file in $MADMARK_HOME/schedule, e.g. $MADMARK_HOME/schedule/example.yaml
2) Add relevant case files that you want to run, e.g. (in example.yaml):
   cases : case_example
   platform : GPDB42
3) In the case file (which should also be in the $MADMARK_HOME/schedule directory), include whichever test cases you want to run, e.g. (in case_example):
  kmeans_baseline_0_0
  kmeans_baseline_0_1
  kmeans_baseline_0_2
  kmeans_baseline_0_3
4) Run 'cd $MADMARK_HOME/bin; python run.py -s example.yaml'

Commit:
Copied testspec/casespec/kmeans.xml to testspec/casespec/kmeans_new.xml, to
track changes compared to old kmeans test cases.

Commit:
Copied (and adapted) existing k-means test cases to new k-means

Commit:
Made distance-function and meann-aggregate parameters schema-qualified. Now
use CTAS workaround to be compatible with older versions of Greenplum.

geeg commented on testspec/metadata/algorithmspec.xml:
```diff
@@ -3673,6 +3879,77 @@
+                    , {src_col_data}         -- expr_point
+                    , (SELECT centroids FROM {table_name}) -- centroids
+                    , {dist_metric}          -- fn_dist
+                    );
```
I think we still need the "DROP TABLE {table_name};" in order to be able to run different tests using the same dataset (otherwise, after using the dataset table_name once, all other tests that use the same dataset ERROR out since the table_name already exists).

fschopp replied:
I think a `teardown` section should be used for that. This is [what I did](https://github.com/fschopp/testsuite/blob/b16151586bd8a7039ec216595efe34a969aef1c9/testspec/casespec/kmeans_new.xml#L2501).

iyerr3 commented on testspec/metadata/algorithmspec.xml:
```diff
@@ -3497,6 +3624,85 @@
+            </input_parameter>
+            <input_parameter>
+                <name>k</name>
+                <type>int</type>
```
I had to change this to 'smallint' to make it work

Commit:
Distance argument to k-means and silhouette function now supplied as a pair.
Added declarations for kmeanspp_seeding and kmeans_random_seeding.

Commit:
Allow argument values that contain double quotes

More complex arguments may contain quotes, e.g.,
ARRAY['madlib.squared_dist_norm2','madlib.dist_norm2']. Previously, quotes did
not pass through the shell invocation but caused errors.
